### PR TITLE
Add support for kube cert signing requests

### PR DESCRIPTION
**Pull request by jandubois**
When the `ca` field for an ssl cert definition is empty, create a kube CSR instead.

If the `-autoApprove` commandline option is set, then also attempt to approve the request.

The secrets generator will wait until all kube CSRs have been approved. Failure to auto-approve are ignored, but any errors while checking the status of pending CSRs will terminate the generator with an error.

Add `append_kube_ca` property to CA cert options. When `true`, the secrets generator will append the `ca.crt` from the kube service account to the CA cert generated for this variable.

Allow template expansion for all variable options. Previously this was hardcoded for the `alternative_names` list only, but now is available anywhere under the `options` key.

In addition this commit adds a `--set` commandline option to specify additional template environment variables.

Unrelated, this commit also removes the `common_name` setting, which is not used anywhere in the cert generation code.

### CODE REVIEW

File: main.go
Comment (mook-as) [inline main.go]: Returning an empty string could be confusing if we ever see this.

File: model/manifest.go
Comment (mook-as) [inline model/manifest.go:80]: We should document what `env` is for in the method doc comment, so that [the documentation](https://godoc.org/github.com/SUSE/scf-secret-generator/model#GetManifest) is reasonable.

File: model/manifest.go
Comment (mook-as) [inline model/manifest.go:101]: I take it that we're discarding the result just to verify the expansion? If yes, document that.

If not, document what we _are_ doing… it's not modifying `raw["variables"]` in place is it? It might be better to just copy in that case…

File: model/manifest.go
Comment (mook-as) [inline model/manifest.go:187]: It would be good to document the path taken to get to this string.

File: secrets/secrets.go
Comment (mook-as) [inline secrets/secrets.go:146]: The documentation doesn't match the function name. (And the function name is rather a mouthful.)

File: ssl/ssl.go
Comment (mook-as) [inline ssl/ssl.go:155]: Is there a point to attempting to sign the certs when stuff is known to be broken? So we can go delete the CSRs? If yes, it may be clearer to do that with `defer` (but then it might make the error reporting harder…)

File: ssl/ssl.go
Comment (mook-as) [inline ssl/ssl.go:337]: Do we need to figure out the procedure to renew the certs? (Probably as a new task)

File: ssl/ssl.go
Comment (mook-as) [inline ssl/ssl.go]: The only possible values here are approved or denied (there are no unrelated conditions in this array), right? We can't have the approval condition later in the condition list?

File: ssl/ssl_test.go
Comment (mook-as) [inline ssl/ssl_test.go]: Use backticks for the string delimiter to skip the need to escape backslashes.  Too bad there isn't `assert.HasPrefix` or something like that…

File: ssl/ssl_test.go
Comment (mook-as) [inline ssl/ssl_test.go]: Shouldn't we test that the content of `kubeca.crt` was inserted (and not, say, two copies of the internal CA)?

**Review by mook-as — CHANGES_REQUESTED** [review]
Mostly fine, just some localized changes.

File: main.go
Comment (jandubois) [reply main.go]: I don't know when we would see this. And I have no idea how to represent a map value here. I'm happy to put in whatever you feel appropriate.

Given that the only caller of this program is a bash script under our control too, I'm not sure it matters much.

File: model/manifest.go
Comment (jandubois) [reply model/manifest.go:101]: > it's not modifying raw["variables"] in place is it? It might be better to just copy in that case…

It is doing just that. Recursively copying larger and larger tree structures in general feels like a bad idea, although in this particular instance the structures are reasonably small.

I did run into some problems when trying to implement everything via copying, but I can't remember anymore (this commit is already 6 weeks old).

I'm happy to add a comment, but not really keen to rewrite this from scratch, given that it works "as documented" (`expandTemplates` docs says that it expands "in-place").

File: ssl/ssl.go
Comment (jandubois) [reply ssl/ssl.go:155]: Yes, this exists only to make sure we delete the outstanding CSRs.

I don't think using `defer` will make this clearer, because it interferes with proper error handling, as you already noted.

However, this now reminds me that I wanted to attempt do delete the CSR before creating it, just in case it somehow escaped proper cleanup in an earlier run.

File: ssl/ssl.go
Comment (jandubois) [reply ssl/ssl.go:337]: The procedure should be the same as for any other certs we generate: secrets rotation.

I would hope that maybe Stratos will eventually have a way to warn an admin about certs that are going to expire soon. Alternatively the admin could run something like `havener` to quickly get the current state.

File: ssl/ssl.go
Comment (jandubois) [reply ssl/ssl.go]: Technically it would be possible to have multiple `Conditions` in the slice, but I don't see how that makes sense.

I guess it really depends on the cert approver (you could have a custom implementation), but given that there are only 2 defined type values, I don't know what we should do. Loop through all conditions and stop at the first approved or denied type?

File: model/manifest.go
Comment (jandubois) [reply model/manifest.go:187]: Yeah, but where do I get the path?

I expect this error to just catch malformed template expressions, and the error should include the full template. So it should be easy enough to find this in the rolemanifest.

File: main.go
Comment (jandubois) [reply main.go]: I tried to return ` fmt.Sprintf("%v", env)`, but that shows a warning when running the tests:

```
./main.go:72:9: Sprintf format %v with arg env causes recursive String method call
```

I think returning the empty string is fine, given that the code will not be executed anyways.

File: model/manifest.go
Comment (jandubois) [reply model/manifest.go:80]: Fixed in 8cf5338

File: model/manifest.go
Comment (jandubois) [reply model/manifest.go:101]: Fixed in 8cf5338

File: secrets/secrets.go
Comment (jandubois) [reply secrets/secrets.go:146]: Fixed in 8cf5338.

The name is indeed rather long, but it follows the same name scheme used for 2 other functions (and interfaces), so I decided to leave it as-is.

File: ssl/ssl_test.go
Comment (jandubois) [reply ssl/ssl_test.go]: Fixed in 8cf5338

File: ssl/ssl_test.go
Comment (jandubois) [reply ssl/ssl_test.go]: Fixed in 8cf5338

File: model/manifest.go
Comment (jandubois) [reply model/manifest.go:80]: Fixed in 8cf5338

File: model/manifest.go
Comment (jandubois) [reply model/manifest.go:101]: I've documented what we are doing, so resolving this one as well. :)

File: main.go
Comment (mook-as) [reply main.go]: Sure, or just some hard coded `<env settings>` string or something; I'm mostly thinking _empty_ string would be confusing (since we wouldn't be able to troubleshoot).

File: model/manifest.go
Comment (mook-as) [reply model/manifest.go:187]: We _could_ pass in the path as we recurse through the object. But it's probably fine for now… it's just annoying later :)

File: ssl/ssl.go
Comment (mook-as) [reply ssl/ssl.go]: That was my thought, yes.

File: ssl/ssl.go
Comment (mook-as) [reply ssl/ssl.go:337]: Okay, we _do_ have it [documented](https://susedoc.github.io/doc-cap/develop/cap-guides/html/cha-cap-certificates.html#sec-cap-secrets-rotation), so we should be fine.

File: main.go
Comment (jandubois) [reply main.go]: Fixed in c605f62

File: ssl/ssl.go
Comment (jandubois) [reply ssl/ssl.go]: Fixed in c605f62

**Review by mook-as — APPROVED** [review]

**Pull request merged**
